File: versionJava/RandomSwapDoubleCycles.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;

public class RandomSwapDoubleCycles implements LLHFunctionInterface{

   private int counter;
   private double score;
   private long timestamp;
   private double f1;
   private Map<LLHFunctionInterface,Double> f2 = new HashMap<>();
   Map<LLHFunctionInterface, Integer> consec_counts = new HashMap<>();
    public RandomSwapDoubleCycles() {
            counter = 0;
            score = 0;
            timestamp = 0;
            f1 = 0;
            
     }
   @Override
    public List<Integer> apply(List<Integer> orgTour) {
        return randomSwapDoubleCycles(orgTour);
    }
    

    private  List<Integer> randomSwapDoubleCycles(List<Integer> orgTour) {
        List<Integer> tour = new ArrayList<>(orgTour);
        List<List<Integer>> reversedSublist = new ArrayList<>();
        for (int i = 0; i < tour.size(); i=i+2) {
           reversedSublist.add(tour.subList(i, i+2));

        }
        List<List<List<Integer>>> squenceList = new ArrayList<>();
        squenceList.add(reversedSublist.subList(0, reversedSublist.size()/2));        
        squenceList.add(reversedSublist.subList(reversedSublist.size()/2, reversedSublist.size()));
        for (List<List<Integer>> list : squenceList) {
            List<Integer> saveIndexSwap = new ArrayList<>();
            for (int i = 0; i < list.size()/2; i++) {
                Random random = new Random();
                int sub1 = random.nextInt(list.size());
                int sub2 = random.nextInt(list.size());
                while(saveIndexSwap.contains(sub1)) {
                    sub1 = random.nextInt(list.size());
                }
                while(true) {
                    sub2 = random.nextInt(list.size());
                    if(sub1 != sub2 && !saveIndexSwap.contains(sub2)) {
                        break;
                    }
                }
                saveIndexSwap.add(sub1);
                saveIndexSwap.add(sub2);
                Collections.swap(list, sub1, sub2);
            }
        }
        List<Integer> result = new ArrayList<>();
        for (List<List<Integer>> list : squenceList) {
            for (List<Integer> list2 : list) {
                result.addAll(list2);
            }
        }
        return result;
    }

    public Integer getConsec_counts(LLHFunctionInterface function) {
        return consec_counts.get(function);
    }
    public void setConsec_counts(LLHFunctionInterface function, Integer count) {
        consec_counts.put(function, count);
    }
    public void addConsec_counts(LLHFunctionInterface function) {
        if (consec_counts.containsKey(function)) {
            consec_counts.put(function, consec_counts.get(function) + 1);
        } else {
            consec_counts.put(function, 1);
        }
    }
    @Override
    public int getCounter() {
       return counter;
    }
    @Override
    public void incrementCounter() {
        counter++;
    }
    @Override
    public void setScore(double score) {
        this.score = score;
    }
    @Override
    public double getScore() {
        return score;
    }
    public void setCounter(int counter) {
        this.counter = counter;
    }
    @Override
    public long getTimestamp() {
        return timestamp;
    }
    @Override
    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
        
    }
    @Override
    public double getF1() {
        return this.f1; 
    }
    @Override
    public void setF1(double f1) {
        this.f1 = f1;
    }
    @Override
    public double getF2(LLHFunctionInterface function) {
        if(f2.containsKey(function)) {
            return f2.get(function);
        } else {
            return 0;
        }
    }
    @Override
    public void addF2(LLHFunctionInterface function, double f2) {
        if(this.f2.containsKey(function)) {
            this.f2.put(function, this.f2.get(function) + f2);
        } else {
            this.f2.put(function, f2);
        }
    }
}

File: tse/RandomSwapDoubleCycles.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class RandomSwapDoubleCycles : ILLHFunctionInterface
{
    private int counter;
    //Số lần được gọi đển của llh này
    private double score;
    //Điểm được gọi đển của llh này
    private long timestamp;
    //Thời gian chạy gần nhất của llh này
    private double f1;
    //Giá trị f1 gần nhất
    private Dictionary<ILLHFunctionInterface, double> f2 = new Dictionary<ILLHFunctionInterface, double>();
    //Danh sách giá trị f2 tương ứng với từng cặp llhs
    private Dictionary<ILLHFunctionInterface, int> consecCounts = new Dictionary<ILLHFunctionInterface, int>();
    //Danh sách số lần được gọi cùng nhau của từng cặp llhs

    public RandomSwapDoubleCycles()
    {        //Khởi tạo các giá trị ban đầu của functions

        counter = 0;
        score = 0;
        timestamp = 0;
        f1 = 0;
    }

    public List<int> Apply(List<int> orgTour)
    {
        return RandomSwapDoubleCyclesFunction(orgTour);
    }

    private List<int> RandomSwapDoubleCyclesFunction(List<int> tour)
    {
        // Random Swap Double Cycle (RSDC): Hoán đổi ngẫu nhiên theo cặp và hoán đổi đoạn tour

        // Chọn ngẫu nhiên các cặp đoạn tour có cùng số thành phố (ít nhất 4).
        // Hoán đổi vị trí của các thành phố trong từng cặp đoạn.
        // Hoán đổi vị trí của hai đoạn tour đã hoán đổi.
            if (tour.Count < 8)
            {
            return tour;
            }

            var random = new Random();

            // Select two pairs of indices (i1, i2) and (j1, j2) randomly within the tour bounds.
            int i1 = random.Next(tour.Count / 2);
            int i2 = i1 + random.Next(1, tour.Count / 2);
            int j1 = random.Next(tour.Count / 2);
            while (j1 == i1 || j1 == i2)
            {
                j1 = random.Next(tour.Count / 2);
            }
            int j2 = j1 + random.Next(1, tour.Count / 2);
            while (j2 == i1 || j2 == i2 || j2 == j1)
            {
                j2 = random.Next(tour.Count / 2);
            }

            // Ensure the selected pairs have the same length.
            if (i2 - i1 != j2 - j1)
            {
            return tour;
            }

            // Swap city positions within each pair.
            var temp1 = tour[i1];
            tour[i1] = tour[i2];
            tour[i2] = temp1;
            temp1 = tour[j1];
            tour[j1] = tour[j2];
            tour[j2] = temp1;

            var subSequence1 = tour.GetRange(i1, i2 - i1 + 1);
            tour.RemoveRange(i1, subSequence1.Count);
            tour.InsertRange(j1, subSequence1);

            return tour;
    }

     public double GetF2(ILLHFunctionInterface function)
    //Input :
    //      function : function cần lấy giá trị F2
    //Output : 
    //      Giá trị f2 tương ứng với function được truyền vào
    {
        if (f2.ContainsKey(function))
        //Nếu function đã tồn tại thì lấy ra giá trị f2 tương ứng nếu không thì trả về 0
        {
            return f2[function];
        }
        else
        {
            return 0;
        }
    }

    public void AddF2(ILLHFunctionInterface function, double f2Value)
    //Set giá trị f2 của function tương ứng 
    //Input :
    //      function : function cần set giá trị 
    //      f2value  : Giá trị f2 tương ứng 
    {
        if (f2.ContainsKey(function))
        {
            f2[function] = f2Value;
        }
        else
        {
            f2[function] = f2Value;
        }
    }

    public int GetConsecCounts(ILLHFunctionInterface function)
    //Lấy số lần được gọi cùng nhau của 2 llhs 
    //Nếu đã tồn tại thì trả về số lần được gọi cùng nhau của 2 này và llh được truyền vào
    //Nếu không thì trả về 0
    {
        if (consecCounts.ContainsKey(function))
        {
            return consecCounts[function];
        }
        else
        {
            return 0;
        }
    }

    public void SetConsecCounts(ILLHFunctionInterface function, int count)
    {
        consecCounts[function] = count;
    }
    
    public void AddConsecCounts(ILLHFunctionInterface function)
    //Tăng số lần được gọi cùng nhau của 2 llh lên 1
    {

        if (consecCounts.ContainsKey(function))
        {
            consecCounts[function]++;
        }
        else
        {
            consecCounts[function] = 1;
        }
    }

    public int GetCounter()
    //Lấy số lần được gọi đến của llh này
    
    {
        return counter;
    }

    public void IncrementCounter()
    //Tăng số lần được gọi của lln này
    {
        counter++;
    }

    public void SetScore(double scoreValue)
    //Set điểm của llh này
    {
        score = scoreValue;
    }

    public double GetScore()
    //Lấy điểm của llh này
    {
        return score;
    }

    public void SetCounter(int counterValue)
    //Set điểm của llh này
    {
        counter = counterValue;
    }

    public long GetTimestamp()
    //Lấy lần thời gian chạy của lần được gọi gần nhất của llhs này
    {
        return timestamp;
    }

    public void SetTimestamp(long timestampValue)
    //Set thời gian chạy của lần được gọi gần nhất của llhs này
    {
        timestamp = timestampValue;
    }

    public double GetF1()
    //Lấy giá trị f1 của llh này
    {
        return f1;
    }

    public void SetF1(double f1Value)
    // Set giá trị f1 của llh này
    {
        f1 = f1Value;
    }

}
